File: SkytilsMod-main/src/main/java/skytils/skytilsmod/features/impl/events/GriffinBurrows.java
package skytils.skytilsmod.features.impl.events;

import com.google.common.collect.ImmutableList;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.EntityPlayerSP;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.entity.Entity;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.util.*;
import net.minecraftforge.client.event.ClientChatReceivedEvent;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.event.world.WorldEvent;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import org.apache.commons.lang3.time.StopWatch;
import skytils.skytilsmod.Skytils;
import skytils.skytilsmod.core.structure.FloatPair;
import skytils.skytilsmod.core.structure.GuiElement;
import skytils.skytilsmod.events.DamageBlockEvent;
import skytils.skytilsmod.utils.APIUtil;
import skytils.skytilsmod.utils.RenderUtil;
import skytils.skytilsmod.utils.SBInfo;
import skytils.skytilsmod.utils.Utils;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;


public class GriffinBurrows {

    public static ArrayList<Burrow> burrows = new ArrayList<>();
    public static ArrayList<BlockPos> dugBurrows = new ArrayList<>();
    public static BlockPos lastDugBurrow = null;
    public static ArrayList<PartialBurrow> partialBurrows = new ArrayList<>();

    public static StopWatch burrowRefreshTimer = new StopWatch();
    public static boolean shouldRefreshBurrows = false;

    private static final Minecraft mc = Minecraft.getMinecraft();

    static {
        new GriffinGuiElement();
    }

    @SubscribeEvent
    public void onTick(TickEvent.ClientTickEvent event) {
        EntityPlayerSP player = mc.thePlayer;
        if (SBInfo.getInstance().getLocation() == null || !SBInfo.getInstance().getLocation().equalsIgnoreCase("hub")) return;
        if (!burrowRefreshTimer.isStarted()) burrowRefreshTimer.start();

        if (burrowRefreshTimer.getTime() >= 60000 || shouldRefreshBurrows) {
            burrowRefreshTimer.reset();
            shouldRefreshBurrows = false;
            if (Skytils.config.showGriffinBurrows && Utils.inSkyblock && player != null) {
                for (int i = 0; i < 8; i++) {
                    ItemStack hotbarItem = player.inventory.getStackInSlot(i);
                    if (hotbarItem == null) continue;
                    if (hotbarItem.getDisplayName().contains("Ancestral Spade")) {
                        player.addChatMessage(new ChatComponentText(EnumChatFormatting.GREEN + "Looking for burrows..."));
                        refreshBurrows();
                        break;
                    }
                }
            }
        }
    }

    @SubscribeEvent(receiveCanceled = true, priority = EventPriority.HIGHEST)
    public void onChat(ClientChatReceivedEvent event) {
        String unformatted = StringUtils.stripControlCodes(event.message.getUnformattedText());
        if (Skytils.config.showGriffinBurrows && (unformatted.contains("You died!") || unformatted.contains("You dug out a Griffin Burrow") || unformatted.contains("You finished the Griffin burrow chain!"))) {
            if (lastDugBurrow != null) {
                dugBurrows.add(lastDugBurrow);
                burrows.removeIf(burrow -> burrow.getBlockPos().equals(lastDugBurrow));
            }
        }
    }

    @SubscribeEvent
    public void onDamageBlock(DamageBlockEvent event) {

        if (mc.theWorld == null || mc.thePlayer == null) return;

        IBlockState blockState = mc.theWorld.getBlockState(event.pos);
        ItemStack item = mc.thePlayer.getHeldItem();

        if (Utils.inSkyblock) {
            if (Skytils.config.showGriffinBurrows && item != null) {
                if (item.getDisplayName().contains("Ancestral Spade") && blockState.getBlock() == Blocks.grass) {
                    if (GriffinBurrows.burrows.stream().anyMatch(burrow -> burrow.getBlockPos().equals(event.pos))) {
                        GriffinBurrows.lastDugBurrow = event.pos;
                    }
                }
            }
        }
    }

    @SubscribeEvent
    public void onWorldRender(RenderWorldLastEvent event) {
        if (Skytils.config.showGriffinBurrows && burrows.size() > 0) {
            List<Burrow> burrows = ImmutableList.copyOf(GriffinBurrows.burrows);
            for (Burrow burrow : burrows) {
                burrow.drawWaypoint(event.partialTicks);
            }
        }
    }

    @SubscribeEvent
    public void onWorldChange(WorldEvent.Load event) {
        burrows.clear();
        shouldRefreshBurrows = true;
    }

    public static void refreshBurrows() {
        new Thread(() -> {
            System.out.println("Finding burrows");
            String uuid = mc.thePlayer.getGameProfile().getId().toString().replaceAll("[\\-]", "");
            String apiKey = Skytils.config.apiKey;
            if (apiKey.length() == 0) {
                mc.thePlayer.addChatMessage(new ChatComponentText("\u00a7c\u00a7lYour API key is required in order to use the burrow feature. \u00a7cPlease set it with /api new or /st setkey <key>"));
                Skytils.config.showGriffinBurrows = false;
                return;
            }

            String latestProfile = APIUtil.getLatestProfileID(uuid, apiKey);
            if (latestProfile == null) return;

            JsonObject profileResponse = APIUtil.getJSONResponse("https://api.hypixel.net/skyblock/profile?profile=" + latestProfile + "&key=" + apiKey);
            if (!profileResponse.get("success").getAsBoolean()) {
                String reason = profileResponse.get("cause").getAsString();
                mc.thePlayer.addChatMessage(new ChatComponentText(EnumChatFormatting.RED + "Failed getting burrows with reason: " + reason));
                return;
            }

            JsonObject playerObject = profileResponse.get("profile").getAsJsonObject().get("members").getAsJsonObject().get(uuid).getAsJsonObject();

            if (!playerObject.has("griffin")) {
                mc.thePlayer.addChatMessage(new ChatComponentText(EnumChatFormatting.RED + "Failed getting burrows with reason: No griffin object."));
                return;
            }

            JsonArray burrowArray = playerObject.get("griffin").getAsJsonObject().get("burrows").getAsJsonArray();

            ArrayList<Burrow> receivedBurrows = new ArrayList<>();
            burrowArray.forEach(jsonElement -> {
                JsonObject burrowObject = jsonElement.getAsJsonObject();
                int x = burrowObject.get("x").getAsInt();
                int y = burrowObject.get("y").getAsInt();
                int z = burrowObject.get("z").getAsInt();
                int type = burrowObject.get("type").getAsInt();
                int tier = burrowObject.get("tier").getAsInt();
                int chain = burrowObject.get("chain").getAsInt();
                Burrow burrow = new Burrow(x, y, z, type, tier, chain);
                receivedBurrows.add(burrow);
            });

            dugBurrows.removeIf(dug -> receivedBurrows.stream().noneMatch(burrow -> burrow.getBlockPos().equals(dug)));
            receivedBurrows.removeIf(burrow -> dugBurrows.contains(burrow.getBlockPos()));

            burrows.clear();
            burrows.addAll(receivedBurrows);
            mc.thePlayer.addChatMessage(new ChatComponentText(EnumChatFormatting.GREEN + "Skytils loaded " + EnumChatFormatting.DARK_GREEN + receivedBurrows.size() + EnumChatFormatting.GREEN + " burrows!"));

        }).start();
    }

    public static class GriffinGuiElement extends GuiElement {

        public GriffinGuiElement() {
            super("Griffin Timer", 1.0F, new FloatPair(100, 10));
            Skytils.GUIMANAGER.registerElement(this);
        }

        public void render() {
            if (SBInfo.getInstance().getLocation() == null || !SBInfo.getInstance().getLocation().equalsIgnoreCase("hub")) return;
            EntityPlayerSP player = mc.thePlayer;
            if (Skytils.config.showGriffinCountdown && Utils.inSkyblock && player != null) {
                for (int i = 0; i < 8; i++) {
                    ItemStack hotbarItem = player.inventory.getStackInSlot(i);
                    if (hotbarItem == null) continue;
                    if (hotbarItem.getDisplayName().contains("Ancestral Spade")) {
                        long diff = Math.round((60_000L - GriffinBurrows.burrowRefreshTimer.getTime()) / 1000L);
                        ScaledResolution sr = new ScaledResolution(mc);
                        float x = this.getPos().getX() * sr.getScaledWidth();
                        float y = this.getPos().getY() * sr.getScaledHeight();

                        GlStateManager.scale(this.getScale(), this.getScale(), 1.0);
                        Minecraft.getMinecraft().fontRendererObj.drawString("Time until refresh: " + diff + "s", x, y, 0xFFFFFF, true);
                        GlStateManager.scale(1/this.getScale(), 1/this.getScale(), 1.0F);
                        break;
                    }
                }
            }
        }

        public boolean getToggled() {
            return Skytils.config.showGriffinBurrows;
        }
    }

    public static class PartialBurrow {
        public int x, y, z;
        public boolean hasFootstep, hasOtherParticle;

        public PartialBurrow(int x, int y, int z, boolean hasFootstep, boolean hasOtherParticle) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.hasFootstep = hasFootstep;
            this.hasOtherParticle = hasOtherParticle;
        }

        public PartialBurrow(Vec3i vec3, boolean hasFootstep, boolean hasOtherParticle) {
            this(vec3.getX(), vec3.getY(), vec3.getZ(), hasFootstep, hasOtherParticle);
        }

        public BlockPos getBlockPos() {
            return new BlockPos(x, y, z);
        }

    }

    public static class Burrow {
        public int x, y, z;
        /**
         * This variable seems to hold whether or not the burrow is the start/empty, a mob, or treasure
         */
        public int type;
        /**
         * This variable holds the Griffin used, -1 means no Griffin, 0 means Common, etc.
         */
        public int tier;
        /**
         * This variable appears to hold what order the burrow is in the chain
         */
        public int chain;

        public Burrow(int x, int y, int z, int type, int tier, int chain) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.type = type;
            this.tier = tier;
            this.chain = chain;
        }

        public BlockPos getBlockPos() {
            return new BlockPos(x, y, z);
        }

        public String getWaypointText() {

            String type = "Burrow";

            switch (this.type) {
                case 0:
                    type = chain == 0 ? EnumChatFormatting.GREEN + "Start" : EnumChatFormatting.WHITE + "Empty";
                    break;
                case 1:
                    type = EnumChatFormatting.RED + "Mob";
                    break;
                case 2:
                case 3:
                    type = EnumChatFormatting.GOLD + "Treasure";
                    break;
            }


            FastTravelLocations closest = null;

            if (Skytils.config.showBurrowFastTravel) {
                double distance = mc.thePlayer.getPosition().distanceSq(getBlockPos());
                for (FastTravelLocations warp : FastTravelLocations.values()) {
                    double warpDistance = getBlockPos().distanceSq(warp.pos);
                    if (warpDistance < distance) {
                        distance = warpDistance;
                        closest = warp;
                    }
                }
            }

            return String.format("%s \u00a7bPosition: %s/4%s", type, this.chain + 1, closest != null ? " " + closest.getNameWithColor() : "");
        }

        public void drawWaypoint(float partialTicks) {

            Entity viewer = Minecraft.getMinecraft().getRenderViewEntity();
            double viewerX = viewer.lastTickPosX + (viewer.posX - viewer.lastTickPosX) * partialTicks;
            double viewerY = viewer.lastTickPosY + (viewer.posY - viewer.lastTickPosY) * partialTicks;
            double viewerZ = viewer.lastTickPosZ + (viewer.posZ - viewer.lastTickPosZ) * partialTicks;

            BlockPos pos = this.getBlockPos();
            double x = pos.getX() - viewerX;
            double y = pos.getY() - viewerY;
            double z = pos.getZ() - viewerZ;
            double distSq = x*x + y*y + z*z;

            GlStateManager.disableDepth();
            GlStateManager.disableCull();
            RenderUtil.drawFilledBoundingBox(new AxisAlignedBB(x, y, z, x + 1, y + 1, z + 1), new Color(173, 216, 230), 1f);
            GlStateManager.disableTexture2D();
            if (distSq > 5*5) RenderUtil.renderBeaconBeam(x, y + 1, z, new Color(173, 216, 230).getRGB(), 1.0f, partialTicks);
            RenderUtil.renderWaypointText(getWaypointText(), getBlockPos().up(5), partialTicks);
            GlStateManager.disableLighting();
            GlStateManager.enableTexture2D();
            GlStateManager.enableDepth();
            GlStateManager.enableCull();
        }

    }

    public enum FastTravelLocations {
        CASTLE(-250, 130, 45),
        CRYPTS(-162, 60, -100),
        DA(91, 74, 173),
        HUB(-3, 70, -70);

        int x, y, z;
        BlockPos pos;
        FastTravelLocations(int x, int y, int z) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.pos = new BlockPos(x, y, z);
        }

        public String getNameWithColor() {
            switch (this) {
                case CASTLE:
                    return EnumChatFormatting.GRAY + "CASTLE";
                case CRYPTS:
                    return EnumChatFormatting.DARK_GREEN + "CRYPTS";
                case DA:
                    return EnumChatFormatting.DARK_PURPLE + "DA";
                case HUB:
                    return EnumChatFormatting.WHITE + "HUB";
                default:
                    return "";
            }
        }
    }


}

File: src/main/kotlin/skytils/skytilsmod/features/impl/events/GriffinBurrows.kt
/*
 * Skytils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2022 Skytils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package skytils.skytilsmod.features.impl.events

import gg.essential.universal.UChat
import gg.essential.universal.UResolution
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.network.play.client.C07PacketPlayerDigging
import net.minecraft.network.play.client.C08PacketPlayerBlockPlacement
import net.minecraft.network.play.server.S2APacketParticles
import net.minecraft.util.AxisAlignedBB
import net.minecraft.util.BlockPos
import net.minecraft.util.EnumParticleTypes
import net.minecraft.util.Vec3i
import net.minecraftforge.client.event.ClientChatReceivedEvent
import net.minecraftforge.client.event.RenderWorldLastEvent
import net.minecraftforge.event.world.WorldEvent
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import net.minecraftforge.fml.common.gameevent.TickEvent.ClientTickEvent
import org.apache.commons.lang3.time.StopWatch
import skytils.hylin.request.HypixelAPIException
import skytils.skytilsmod.Skytils
import skytils.skytilsmod.Skytils.Companion.mc
import skytils.skytilsmod.core.structure.FloatPair
import skytils.skytilsmod.core.structure.GuiElement
import skytils.skytilsmod.events.impl.MainReceivePacketEvent
import skytils.skytilsmod.events.impl.PacketEvent
import skytils.skytilsmod.utils.*
import skytils.skytilsmod.utils.graphics.ScreenRenderer
import skytils.skytilsmod.utils.graphics.SmartFontRenderer
import skytils.skytilsmod.utils.graphics.SmartFontRenderer.TextAlignment
import skytils.skytilsmod.utils.graphics.colors.CommonColors
import java.awt.Color
import java.util.concurrent.Future
import kotlin.math.roundToInt

object GriffinBurrows {
    init {
        GriffinGuiElement()
    }

    var lastManualRefresh = 0L

    val burrows = hashMapOf<BlockPos, Burrow>()
    val dugBurrows = hashSetOf<BlockPos>()
    var lastDugBurrow: BlockPos? = null
    val particleBurrows = hashMapOf<BlockPos, ParticleBurrow>()
    var lastDugParticleBurrow: BlockPos? = null
    val burrowRefreshTimer = StopWatch()
    var shouldRefreshBurrows = false

    var hasSpadeInHotbar = false

    fun refreshBurrows(): Future<*> {
        return Skytils.threadPool.submit {
            try {
                println("Finding burrows")
                val uuid = mc.thePlayer.gameProfile.id
                val apiKey = Skytils.config.apiKey
                if (apiKey.isBlank()) {
                    UChat.chat("§c§lYour API key is required in order to use the burrow feature. §cPlease set it with /api new or /st setkey <key>")
                    Skytils.config.showGriffinBurrows = false
                    return@submit
                }
                val profileData =
                    Skytils.hylinAPI.getLatestSkyblockProfileForMemberSync(uuid)
                if (profileData == null) {
                    UChat.chat("§c§lUnable to find your Skyblock Profile!")
                    return@submit
                }
                val receivedBurrows = profileData.griffin.burrows.associateTo(hashMapOf()) {
                    val b = Burrow(it.x, it.y, it.z, it.type, it.tier, it.chain)
                    b.blockPos to b
                }
                Utils.checkThreadAndQueue {
                    dugBurrows.removeAll {
                        !receivedBurrows.containsKey(it)
                    }
                    particleBurrows.keys.removeAll {
                        receivedBurrows.containsKey(it)
                    }
                    val dupes = receivedBurrows.filterTo(hashMapOf()) { (bpos, _) ->
                        dugBurrows.contains(bpos) || particleBurrows[bpos]?.dug == true
                    }.also { receivedBurrows.entries.removeAll(it.entries) }
                    burrows.clear()
                    burrows.putAll(receivedBurrows)
                    particleBurrows.clear()
                    if (receivedBurrows.size == 0) {
                        if (dupes.isEmpty()) UChat.chat("§cSkytils failed to load griffin burrows. Try manually digging a burrow and switching hubs.") else UChat.chat(
                            "§cSkytils was unable to load fresh burrows. Please wait for the API refresh or switch hubs."
                        )
                    } else UChat.chat("§aSkytils loaded §2${receivedBurrows.size}§a burrows!")
                }
            } catch (apiException: HypixelAPIException) {
                UChat.chat("§cFailed to get burrows with reason: ${apiException.message}")
            } catch (e: Exception) {
                UChat.chat("§cSkytils ran into a fatal error whilst fetching burrows, please report this on our Discord. ${e::class.simpleName}: ${e.message}")
                e.printStackTrace()
            }
        }
    }


    @SubscribeEvent
    fun onTick(event: ClientTickEvent) {
        val player = mc.thePlayer
        if (event.phase != TickEvent.Phase.START) return
        hasSpadeInHotbar = player != null && Utils.inSkyblock && (0..7).any {
            player.inventory.getStackInSlot(it).isSpade
        }
        if (!Utils.inSkyblock || player == null || !Skytils.config.showGriffinBurrows || SBInfo.mode != SkyblockIsland.Hub.mode) return
        if (!burrowRefreshTimer.isStarted) burrowRefreshTimer.start()
        if ((burrowRefreshTimer.time >= 60_000L || shouldRefreshBurrows)) {
            burrowRefreshTimer.reset()
            shouldRefreshBurrows = false
            if (hasSpadeInHotbar) {
                UChat.chat("§aSkytils is looking for burrows...");
                refreshBurrows()
            }
        }
    }

    @SubscribeEvent(receiveCanceled = true, priority = EventPriority.HIGHEST)
    fun onChat(event: ClientChatReceivedEvent) {
        val unformatted = event.message.unformattedText.stripControlCodes()
        if (Skytils.config.showGriffinBurrows &&
            (unformatted.startsWith("You died") ||
                    unformatted.startsWith("You dug out a Griffin Burrow! (") ||
                    unformatted == "You finished the Griffin burrow chain! (4/4)")
        ) {
            if (lastDugBurrow != null) {
                dugBurrows.add(lastDugBurrow!!)
                burrows.remove(lastDugBurrow!!)
                lastDugBurrow = null
            }
            if (lastDugParticleBurrow != null) {
                val particleBurrow =
                    particleBurrows[lastDugParticleBurrow] ?: return
                particleBurrow.dug = true
                dugBurrows.add(particleBurrow.blockPos)
                particleBurrows.remove(particleBurrow.blockPos)
                lastDugParticleBurrow = null
            }
        }
    }

    @SubscribeEvent
    fun onSendPacket(event: PacketEvent.SendEvent) {
        if (!Utils.inSkyblock || !Skytils.config.showGriffinBurrows || mc.theWorld == null || mc.thePlayer == null) return
        val pos =
            when {
                event.packet is C07PacketPlayerDigging && event.packet.status == C07PacketPlayerDigging.Action.START_DESTROY_BLOCK -> {
                    event.packet.position
                }
                event.packet is C08PacketPlayerBlockPlacement && event.packet.stack != null -> event.packet.position
                else -> return
            }
        if (mc.thePlayer.heldItem?.isSpade != true || mc.theWorld.getBlockState(pos).block !== Blocks.grass) return
        lastDugBurrow = burrows[pos]?.blockPos ?: lastDugBurrow
        lastDugParticleBurrow = particleBurrows[pos]?.blockPos ?: lastDugParticleBurrow
    }

    @SubscribeEvent
    fun onWorldRender(event: RenderWorldLastEvent) {
        if (Skytils.config.showGriffinBurrows) {
            for (burrow in burrows.values) {
                burrow.drawWaypoint(event.partialTicks)
            }
            if (Skytils.config.particleBurrows) {
                for (pb in particleBurrows.values) {
                    if (pb.hasEnchant && pb.hasFootstep && pb.type != -1) {
                        pb.drawWaypoint(event.partialTicks)
                    }
                }
            }
        }
    }

    @SubscribeEvent
    fun onWorldChange(event: WorldEvent.Load) {
        burrows.clear()
        particleBurrows.clear()
        shouldRefreshBurrows = true
    }

    class GriffinGuiElement : GuiElement("Griffin Timer", FloatPair(100, 10)) {
        override fun render() {
            if (SBInfo.mode != SkyblockIsland.Hub.mode) return
            val player = mc.thePlayer
            if (toggled && Utils.inSkyblock && player != null && hasSpadeInHotbar) {
                val diff = ((60_000L - burrowRefreshTimer.time) / 1000L).toFloat().roundToInt().toLong()
                val sr = UResolution
                val leftAlign = actualX < sr.scaledWidth / 2f
                val alignment = if (leftAlign) TextAlignment.LEFT_RIGHT else TextAlignment.RIGHT_LEFT
                ScreenRenderer.fontRenderer.drawString(
                    "Time until refresh: " + diff + "s",
                    if (leftAlign) 0f else actualWidth,
                    0f,
                    CommonColors.WHITE,
                    alignment,
                    SmartFontRenderer.TextShadow.NORMAL
                )
            }
        }

        override fun demoRender() {
            ScreenRenderer.fontRenderer.drawString(
                "Time until refresh: 10s",
                0f,
                0f,
                CommonColors.WHITE,
                TextAlignment.LEFT_RIGHT,
                SmartFontRenderer.TextShadow.NORMAL
            )
        }

        override val height: Int
            get() = ScreenRenderer.fontRenderer.FONT_HEIGHT
        override val width: Int
            get() = ScreenRenderer.fontRenderer.getStringWidth("Time until refresh: 10s")
        override val toggled: Boolean
            get() = Skytils.config.showGriffinBurrows && Skytils.config.showGriffinCountdown

        init {
            Skytils.guiManager.registerElement(this)
        }
    }

    @SubscribeEvent
    fun onReceivePacket(event: MainReceivePacketEvent<*, *>) {
        if (!Utils.inSkyblock) return
        if (Skytils.config.showGriffinBurrows && Skytils.config.particleBurrows && event.packet is S2APacketParticles) {
            if (SBInfo.mode != SkyblockIsland.Hub.mode) return
            event.packet.apply {
                val pos = BlockPos(x, y, z).down()
                val footstepFilter =
                    type == EnumParticleTypes.FOOTSTEP && count == 1 && speed == 0.0f && xOffset == 0.05f && yOffset == 0.0f && zOffset == 0.05f
                val enchantFilter =
                    type == EnumParticleTypes.ENCHANTMENT_TABLE && count == 5 && speed == 0.05f && xOffset == 0.5f && yOffset == 0.4f && zOffset == 0.5f
                val startFilter =
                    type == EnumParticleTypes.CRIT_MAGIC && count == 4 && speed == 0.01f && xOffset == 0.5f && yOffset == 0.1f && zOffset == 0.5f
                val mobFilter =
                    type == EnumParticleTypes.CRIT && count == 3 && speed == 0.01f && xOffset == 0.5f && yOffset == 0.1f && zOffset == 0.5f
                val treasureFilter =
                    type == EnumParticleTypes.DRIP_LAVA && count == 2 && speed == 0.01f && xOffset == 0.35f && yOffset == 0.1f && zOffset == 0.35f
                if (isLongDistance && (footstepFilter || enchantFilter || startFilter || mobFilter || treasureFilter)) {
                    if (!burrows.containsKey(pos) && !dugBurrows.contains(pos)) {
                        if (burrows.keys.any { it.distanceSq(x, y, z) < 4 }) return
                        val burrow = particleBurrows.getOrPut(pos) {
                            ParticleBurrow(pos, hasFootstep = false, hasEnchant = false, type = -1)
                        }

                        if (!burrow.hasFootstep && footstepFilter) {
                            burrow.hasFootstep = true
                        } else if (!burrow.hasEnchant && enchantFilter) {
                            burrow.hasEnchant = true
                        } else if (burrow.type == -1 && type != EnumParticleTypes.FOOTSTEP && type != EnumParticleTypes.ENCHANTMENT_TABLE) {
                            when {
                                startFilter -> burrow.type = 0
                                mobFilter -> burrow.type = 1
                                treasureFilter -> burrow.type = 2
                            }
                        }
                    }
                }
            }
        }
    }

    abstract class Diggable {
        abstract val x: Int
        abstract val y: Int
        abstract val z: Int
        abstract var type: Int
        val blockPos: BlockPos by lazy {
            BlockPos(x, y, z)
        }

        protected abstract val waypointText: String
        protected abstract val color: Color
        fun drawWaypoint(partialTicks: Float) {
            val (viewerX, viewerY, viewerZ) = RenderUtil.getViewerPos(partialTicks)
            val pos = blockPos
            val x = pos.x - viewerX
            val y = pos.y - viewerY
            val z = pos.z - viewerZ
            val distSq = x * x + y * y + z * z
            GlStateManager.disableDepth()
            GlStateManager.disableCull()
            RenderUtil.drawFilledBoundingBox(
                AxisAlignedBB(x, y, z, x + 1, y + 1, z + 1).expandBlock(),
                this.color,
                (0.1f + 0.005f * distSq.toFloat()).coerceAtLeast(0.2f)
            )
            GlStateManager.disableTexture2D()
            if (distSq > 5 * 5) RenderUtil.renderBeaconBeam(x, y + 1, z, this.color.rgb, 1.0f, partialTicks)
            RenderUtil.renderWaypointText(
                waypointText,
                blockPos.x + 0.5,
                blockPos.y + 5.0,
                blockPos.z + 0.5,
                partialTicks
            )
            GlStateManager.disableLighting()
            GlStateManager.enableTexture2D()
            GlStateManager.enableDepth()
            GlStateManager.enableCull()
        }
    }

    data class ParticleBurrow(
        override val x: Int,
        override val y: Int,
        override val z: Int,
        var hasFootstep: Boolean,
        var hasEnchant: Boolean,
        override var type: Int
    ) : Diggable() {
        var dug = false

        constructor(vec3: Vec3i, hasFootstep: Boolean, hasEnchant: Boolean, type: Int) : this(
            vec3.x,
            vec3.y,
            vec3.z,
            hasFootstep,
            hasEnchant,
            type
        )

        override val waypointText: String
            get() {
                var type = "Burrow"
                when (this.type) {
                    0 -> type = "§aStart"
                    1 -> type = "§cMob"
                    2 -> type = "§6Treasure"
                }
                return "$type §a(Particle)"
            }
        override val color: Color
            get() {
                return when (this.type) {
                    0 -> Skytils.config.emptyBurrowColor
                    1 -> Skytils.config.mobBurrowColor
                    2 -> Skytils.config.treasureBurrowColor
                    else -> Color.WHITE
                }
            }
    }

    data class Burrow(
        override val x: Int, override val y: Int, override val z: Int,
        /**
         * This variable seems to hold whether or not the burrow is the start/empty, a mob, or treasure
         */
        override var type: Int,
        /**
         * This variable holds the Griffin used, -1 means no Griffin, 0 means Common, etc.
         */
        val tier: Int,
        /**
         * This variable appears to hold what order the burrow is in the chain
         */
        private val chain: Int
    ) : Diggable() {
        override val waypointText: String
            get() {
                var type = "Burrow"
                when (this.type) {
                    0 -> type =
                        if (chain == 0) "§aStart" else "§fEmpty"
                    1 -> type = "§cMob"
                    2, 3 -> type = "§6Treasure"
                }
                var closest: FastTravelLocations? = null
                var distance = mc.thePlayer.position.distanceSq(blockPos)
                for (warp in FastTravelLocations.values()) {
                    if (!warp.toggled) continue
                    val warpDistance = blockPos.distanceSq(warp.pos)
                    if (warpDistance < distance) {
                        distance = warpDistance
                        closest = warp
                    }
                }
                return "$type §bPosition: ${chain + 1}/4${
                    if (closest != null) " ${closest.nameWithColor}" else ""
                }"
            }

        override val color: Color
            get() {
                return when (this.type) {
                    0 -> Skytils.config.emptyBurrowColor
                    1 -> Skytils.config.mobBurrowColor
                    2, 3 -> Skytils.config.treasureBurrowColor
                    else -> Color.WHITE
                }
            }
    }

    enum class FastTravelLocations(val nameWithColor: String, val pos: BlockPos) {
        CASTLE("§7CASTLE", BlockPos(-250, 130, 45)),
        CRYPTS("§2CRYPTS", BlockPos(-162, 60, -100)),
        DA("§5DA", BlockPos(91, 74, 173)),
        HUB("§fHUB", BlockPos(-3, 70, -70)),
        MUSEUM("§bMUSEUM", BlockPos(-76, 76, 80));

        val toggled: Boolean
            get() = when (this) {
                CASTLE -> Skytils.config.burrowCastleFastTravel
                CRYPTS -> Skytils.config.burrowCryptsFastTravel
                DA -> Skytils.config.burrowDarkAuctionFastTravel
                HUB -> Skytils.config.burrowHubFastTravel
                MUSEUM -> Skytils.config.burrowMuseumFastTravel
            }
    }

    private val ItemStack?.isSpade
        get() = ItemUtil.getSkyBlockItemID(this) == "ANCESTRAL_SPADE"
}